[smt2_convt] Fix invalid and NULL pointer handling

SaswatPadhi opened the pull request:
Fixes #6022.

Several SMT2 conversion functions used to crash on `NULL` pointers and other kinds of invalid pointers. This change fixes the handling of NULL pointers at several locations.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- NA ~~Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.~~
- NA ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- NA ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- NA ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

SaswatPadhi committed:
[smt2_convt] Fix arithmetic on invalid pointers

CBMC crashed on SMT2 conversion of pointer arithmetic on NULL and other kinds
of invalid pointers.
This change fixes the encoding of pointer arithmetic on invalid pointers.

tautschnig commented on regression/cbmc/memset_null_02/main.c:
```diff
@@ -0,0 +1,8 @@
+#include <string.h>
```
Do we actually need all these tests, or would one of them do the trick? Asking differently: does some variation of the fix make one of them, but not all of them fail/pass?

SaswatPadhi replied:
I removed `memset_null_01` since it's subsumed (during symex) by `memset_null_02`, but the other test was necessary and now I have split the tests into two different commits, as per your suggestion.

tautschnig commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -3163,7 +3163,13 @@ void smt2_convt::convert_plus(const plus_exprt &expr)
+        // so ignore the NULL and simply convert the offset
+        convert_expr(i);
+        return;
+      }
```
I'd do as is done in `bv_pointers.cpp`: even before calling `pointer_offset_size` test for `expr.type().subtype().id() == ID_emtpy` and set `element_size` to `1` in that case.

SaswatPadhi replied:
Fixed.

tautschnig commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -4248,6 +4254,12 @@ void smt2_convt::set_to(const exprt &expr, bool value)
+        {
+          // ignore converting invalid pointers with empty type
+          return;
+        }
```
Two comments here:
1) I think your commit does fix two different problems, and those should be covered by two separate commits (with each of them having their particular set of tests).
2) I'd test for `equal_expr.lhs().type().id() == ID_empty` right after line 4244, there is no need for going through several branches before ending up in this case.

SaswatPadhi replied:
Thanks, fixed both as per your suggestions.

SaswatPadhi committed:
[smt_convt] Fix comparison of invalid pointers

The SMT conversion routines crashed on encoding of comparisons over invalid
pointers.
This change fixes this behavior by ignoring comparisons against invalid
pointers.

tautschnig commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -4247,6 +4247,11 @@ void smt2_convt::set_to(const exprt &expr, bool value)
     const equal_exprt &equal_expr=to_equal_expr(expr);
+    if(equal_expr.lhs().type().id() == ID_empty)
+    {
+      // ignore equality checking on pointers with empty type
```
This comment isn't accurate: I'd suggest "ignore equality checking over expressions with empty (void) type"

SaswatPadhi replied:
Thanks, fixed.
